### PR TITLE
rest: improve type parsing

### PR DESCRIPTION
**Pull request by abergou**
Closes https://github.com/rr-/docstring_parser/issues/26

### CODE REVIEW

**Comment by abergou** [comment]
@rr- do you perchance have time to review?

File: docstring_parser/tests/test_rest.py [hunk 1]
```diff
@@ -298,6 +298,16 @@ def test_params() -> None:
...
+    docstring = parse(
+        """
+        Short description
+
+        :param a: description a
+        :type a: int
+        :param int b: description b
+        """
+    )
+
```
Comment (rr-) [inline docstring_parser/tests/test_rest.py:310]: Missing assertions

File: docstring_parser/tests/test_rest.py [hunk 2]
```diff
@@ -338,6 +348,21 @@ def test_returns() -> None:
...
+        :rtype: int
+        """
+    )
+    assert docstring.returns is not None
+    assert docstring.returns.type_name == "int"
+    assert docstring.returns.description == "description"
+    assert not docstring.returns.is_generator
+    assert docstring.many_returns is not None
+    assert len(docstring.many_returns) == 1
+    assert docstring.many_returns[0] == docstring.returns
```
Comment (rr-) [inline docstring_parser/tests/test_rest.py:364]: Can be shortened to `assert docstring.many_returns == [docstring.returns]`

File: docstring_parser/rest.py [hunk 3]
```diff
@@ -139,11 +141,25 @@ def parse(text: str) -> Docstring:
...
+        elif len(args) in [1, 2] and args[0] == "rtype":
+            rtypes[None if len(args) == 1 else args[1]] = desc
+        else:
+            ret.meta.append(_build_meta(args, desc))
+    for meta in ret.meta:
+        if isinstance(meta, DocstringParam):
+            meta.type_name = meta.type_name or types.get(meta.arg_name, None)
+        elif isinstance(meta, DocstringReturns):
+            meta.type_name = meta.type_name or rtypes.get(
+                meta.return_name, None
```
Comment (rr-) [inline docstring_parser/rest.py:161]: No need for explicit `, None`

File: docstring_parser/rest.py [hunk 4]
```diff
@@ -139,11 +141,25 @@ def parse(text: str) -> Docstring:
...
+        # Add special handling for :type a: typename
+        if len(args) == 2 and args[0] == "type":
+            types[args[1]] = desc
+        elif len(args) in [1, 2] and args[0] == "rtype":
+            rtypes[None if len(args) == 1 else args[1]] = desc
+        else:
+            ret.meta.append(_build_meta(args, desc))
+    for meta in ret.meta:
+        if isinstance(meta, DocstringParam):
+            meta.type_name = meta.type_name or types.get(meta.arg_name, None)
```
Comment (rr-) [inline docstring_parser/rest.py:158]: No need for explicit `, None`

File: docstring_parser/rest.py [hunk 5]
```diff
@@ -139,11 +141,25 @@ def parse(text: str) -> Docstring:
...
 
-        ret.meta.append(_build_meta(args, desc))
+        # Add special handling for :type a: typename
+        if len(args) == 2 and args[0] == "type":
+            types[args[1]] = desc
+        elif len(args) in [1, 2] and args[0] == "rtype":
+            rtypes[None if len(args) == 1 else args[1]] = desc
+        else:
+            ret.meta.append(_build_meta(args, desc))
+    for meta in ret.meta:
```
Comment (rr-) [inline docstring_parser/rest.py:156]: Let's add a blank line before the for loop for clarity that this is postprocessing

**Review by rr- — CHANGES_REQUESTED** [review]
LGTM apart from cosmetic issues

File: docstring_parser/tests/test_rest.py (hunk 1 above)
Comment (abergou) [reply docstring_parser/tests/test_rest.py:316]: fixed.

File: docstring_parser/tests/test_rest.py (hunk 2 above)
Comment (abergou) [reply docstring_parser/tests/test_rest.py]: Fixed

File: docstring_parser/rest.py (hunk 3 above)
Comment (abergou) [reply docstring_parser/rest.py]: Fixed

File: docstring_parser/rest.py (hunk 4 above)
Comment (abergou) [reply docstring_parser/rest.py]: Fixed

File: docstring_parser/rest.py (hunk 5 above)
Comment (abergou) [reply docstring_parser/rest.py:157]: Fixed.

File: docstring_parser/tests/test_rest.py [hunk 6]
```diff
@@ -338,6 +354,21 @@ def test_returns() -> None:
...
+        :returns: description
+        :rtype: int
+        """
+    )
+    assert docstring.returns is not None
+    assert docstring.returns.type_name == "int"
+    assert docstring.returns.description == "description"
+    assert not docstring.returns.is_generator
+    assert docstring.many_returns is not None
+    assert len(docstring.many_returns) == 1
```
Comment (rr-) [inline docstring_parser/tests/test_rest.py:369]: These two assertions are now pointless with the one below.

File: docstring_parser/rest.py [hunk 7]
```diff
@@ -197,15 +212,30 @@ def process_desc(desc: T.Optional[str]) -> str:
...
+            if rendering_style == RenderingStyle.EXPANDED:
+                text = f":{key}:"
+                text += process_desc(meta.description)
+                parts.append(text)
+                if type_text:
+                    parts.append(f":rtype:{type_text}")
+            else:
+                text = f":{key}{type_text}:"
+                text += process_desc(meta.description)
+                parts.append(text)
```
Comment (rr-) [inline docstring_parser/rest.py:238]: Now this part is missing tests.

**Review by rr- — CHANGES_REQUESTED** [review]

**Review by rr- — APPROVED** [review]
Thank you

**Pull request merged**